testcases/expand.go

```diff
@@ -124,6 +125,14 @@ func applyQueryTweaks(tc *comparer.TestCase, tweaks []*config.QueryTweak) *compa
 			resTC.Start = resTC.Start.Truncate(resTC.Resolution)
 			resTC.End = resTC.End.Truncate(resTC.Resolution)
 		}
+		if len(t.QueryStringReplacements.Remove) > 0 {
+			for _, e := range t.QueryStringReplacements.Exemptions {
+				if strings.HasPrefix(resTC.Query, e) {
```

Comment:
Hmm, I think this query tweak implementation is ok as a local stop-gap for your testing, but it's a bit brittle and special-cased for putting it like this in the upstream tester tool. Especially as it only checks the prefix of queries, whereas functions may appear anywhere within a test query string, and the set of test queries is subject to change and expansion over time (so e.g. `rate()` will appear deeper within queries as well, or similar-looking strings will not actually be function calls).

I see two options as nicer alternatives, each of which I'm happy to take on / help with:

* A) Change the set of test queries to not look at absolute values of counters in the first place, use gauges instead. With the exception of a single test case that explicitly tests whether the absolute value of a counter can be retrieved. The main reason why I'm currently using counter metrics in many of the non-`rate()`-y test cases is that the example counter metrics that are exposed by the demo application have a good set of labels to experiment on. If we changed the test dataset to have equally useful gauges for testing, we could just switch those cases over (no replacement magic needed anymore then), and I would even see that as an overall improvement to the test cases to not query so much for absolute values of counters (although for PromQL-compliant implementations, that shouldn't matter).
* B) Add a more robust and general query replacement mechanism that does a full parse of the expression (https://github.com/prometheus/prometheus/blob/63be30dceed942ac53c0cc6d8637e6563598cd87/promql/parser/parse.go#L105), followed by an iteration through the query nodes to apply tweaks in a robust fashion (https://github.com/prometheus/prometheus/blob/63be30dceed942ac53c0cc6d8637e6563598cd87/promql/parser/ast.go#L259-L306), then re-serialize the expression. This is easier than it sounds :) Here is an example of someone parsing an expression and then doing that kind of iteration to enforce the presence of a certain set of labels on selectors: https://github.com/prometheus-community/prom-label-proxy/blob/0211f3ff373505b4bd3368f8c3f9a1962580578b/injectproxy/routes.go#L120-L142. **But the challenge I see this this option is to provide a general-enough way of specifying what to mutate where in the tree, because other people will have similar, but different needs, such as enforcing a label selector, or such.**

I think both alternatives could be actual enrichments to the tester tool, but only one is needed for NR's purposes. 

The first one would be easiest, let me know if you're happy with that and I can get right on that (I might do so in any case).

Reply:
@juliusv I think your solution A sounds like a good plan.  Can you go ahead and implement that?